support job prolog/epilog commands

grondo opened the pull request:
This PR is a WIP implementation of a jobtap plugin and a new command which together support running traditional job prolog/epilogs when the plugin is loaded and configured.

Tests still need to be added, but before that I wanted to ensure the general approach is acceptable.

This PR adds a new `perilog.so` plugin (for lack of a better name) which, when loaded and configured, supports running a configurable command on rank 0 as prolog and/or epilog actions. 

The PR also adds a new helper script `flux-perilog-run.py` which when configured as the prolog or epilog command, supports running a set of scripts from a directory (default `/etc/flux/system/prolog.d` and `/etc/flux/system/epilog.d/`) and optionally executing a command across all ranks of the starting/completing job. This last feature is meant as a stopgap until we have true per-rank prolog/epilog supported by the execution system.

The `perilog.so` plugin reads configuration from the `[job-manager.prolog]` and `[job-manager.epilog]` tables, which currently support a `command` key only, e.g. to run the per-rank prolog/epilog for every job the following configuration would work:

```toml
[job-manager.prolog]
command = [ "flux", "perilog-run", "prolog", "-e/usr/libexec/flux/flux-imp,run,prolog" ]
[job-manager.epilog]
command = [ "flux", "perilog-run", "epilog", "-e/usr/libexec/flux/flux-imp,run,epilog" ]
```

Random notes:
- When running with `-e, --exec-per-rank=CMD[,ARGS...]`, the `flux perilog-run` script will automatically drain any ranks on which the prolog/epilog fails with a corresponding reason.
- If a prolog fails, a job exception is raised before the `prolog-finish` event so that the job is not started
- The job epilog is triggered by a `finish` event, not entry to the CLEANUP state. This means the epilog is only run if a job was started (Hm, I'm rethinking this one now)
- The `flux perilog-run` command will probably go away after the job-exec system incorporates its own prolog/epilog support

grondo commented:
I'm not entirely happy with this solution, but it does at least allow prolog and epilog to be configured and used in a system instance for now. Down the road, there still could be a use case for "job manager prolog/epilog" (i.e. a command that runs once per job on rank 0), so the effort here won't be completely wasted.

To enable the traditional run-as-root job prolog/epilog requires a few steps now:

 1. setup the IMP `[run.prolog]` and `[run.epilog]` sections to allow `flux` user and configure a command:
  ```toml
  [run.prolog]
  allowed-users = [ "flux" ]
  path = "/etc/flux/prolog"
 
  [run.epilog]
  allowed-users = [ "flux" ]
  path = "/etc/flux/system/epilog"
  ```
 2. Load the `perilog.so` plugin by default to support job manager prolog/epilog
  ```toml
  [job-manager]
  plugins = [
    { load = "perilog.so" }
  ]
  ```
 3. Configure the `perilog.so` prolog/epilog to run the `flux perilog-run` helper with `-e, --exec-per-rank` option:
  ```toml
  [job-manager.prolog]
  command = [ 
    "flux", "perilog-run", "prolog", "-e", "/usr/libexec/flux/flux-imp,run,prolog"
  ]
  [job-manager.epilog]
  command = [ 
    "flux", "perilog-run", "epilog", "-e", "/usr/libexec/flux/flux-imp,run,epilog"
  ]
  ```

grondo commented:
And a demonstration:
```console
grondo@asp:~/git/f.git$ cat conf/test.toml
[job-manager]
plugins = [ { load = "perilog.so" } ]

[job-manager.prolog]
command = [
  "flux", "perilog-run", "prolog", "-d", "/tmp/test.d", "-e", "sleep,1"
]

[job-manager.epilog]
command = [ "flux", "perilog-run", "epilog" ]

grondo@asp:~/git/f.git$ src/cmd/flux start -s 4 -o --config-path=conf "flux mini run -n4 -N4 --flags=debug -vv hostname"
jobid: ƒobgRjd
0.000s: job.submit {"userid":1000,"urgency":16,"flags":2}
0.013s: job.depend
0.013s: job.priority {"priority":16}
0.013s: job.debug.alloc-request
0.015s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-3]/core0"}}}
0.015s: job.prolog-start {"description":"job-manager.prolog"}
1.285s: job.prolog-finish {"description":"job-manager.prolog","status":0}
1.285s: job.debug.start-request
1.294s: job.start
asp
asp
asp
asp
1.352s: job.finish {"status":0}
1.353s: job.epilog-start {"description":"job-manager.epilog"}
1.357s: job.release {"ranks":"all","final":true}
1.572s: job.epilog-finish {"description":"job-manager.epilog","status":0}
1.572s: job.debug.free-request
1.572s: job.free
1.572s: job.clean
```

grondo commented:
I've largely reworked the `perilog` plugin to fix a potential for a crash when the plugin is removed.

It is still dangerous to unload the plugin unless the queue is idle -- if any prolog or epilog processes are still outstanding, then unloading the plugin will result  in a dropped `prolog-finish` or `epilog-finish` event and the job will be stuck.

I've added a sharness test now, so I'll remove WIP and this is ready for a review. Apologies in advance for any messiness since the plugin has been rewritten a couple times now!

garlick commented on t/t2274-manager-perilog.t:
```diff
@@ -0,0 +1,190 @@
+#!/bin/sh
+
+test_description='Test porlog/epilog jobtap functionality'
```
s/porlog/prolog/

garlick commented on src/modules/job-manager/plugins/perilog.c:
```diff
@@ -0,0 +1,513 @@
+ *     command = [ "command", "arg1", "arg2" ]
+ *
+ *  - The queue should be idle before unloading/reloading this
+ *     plugin. Otherwise jobs may become stuck becuase a prolog
```
s/becuase/because/

garlick commented on src/modules/job-manager/plugins/perilog.c:
```diff
@@ -0,0 +1,513 @@
+
+static void free_config (struct perilog_conf *conf)
+{
+    if (conf->prolog_cmd)
+        flux_cmd_destroy (conf->prolog_cmd);
+    if (conf->epilog_cmd)
+        flux_cmd_destroy (conf->epilog_cmd);
+    if (conf->processes)
+        zhashx_destroy (&conf->processes);
```
test for non-NULL is not needed for any of these

garlick commented on src/modules/job-manager/plugins/perilog.c:
```diff
@@ -0,0 +1,513 @@
+                            "kill-timeout", &conf->prolog_kill_timeout,
+                          "epilog",
+                            "command", &epilog) < 0) {
+        flux_log_error (h, "prolog/epilog configuration error: %s",
```
`flux_log (h, LOG_ERR, ...` might be more appropriate here (error string appended by `flux_log_error()` is likely redundant with error.errbuf)

garlick commented on src/modules/job-manager/plugins/perilog.c:
```diff
@@ -0,0 +1,513 @@
+    }
+
+    rc = 0;
+done:
```
Remove label and just return -1 from each error path if there is no cleanup to do?

(is sp leaked on `perilog_proc_create()` failure?)

grondo replied:
> (is sp leaked on perilog_proc_create() failure?)

Yes, thanks!

garlick commented on src/cmd/Makefile.am:
```diff
@@ -71,7 +71,8 @@ dist_fluxcmd_SCRIPTS = \
 	flux-admin.py \
 	flux-jobtap.py \
 	flux-job-validator.py \
```
commit message typo in 99eaddd5a52

s/accrossa job's/accross a job's/

garlick commented on t/sharness.d/flux-sharness.sh:
```diff
@@ -137,7 +137,7 @@ remove_trashdir_wrapper() {
 #  Reinvoke a test file under a flux instance
 #
-#  Usage: test_under_flux <size> [personality]
+#  Usage: test_under_flux <size> [personality] [flux-start-options]
```
Oh duh, so obvious and useful!

garlick reviewed (commented):
Noticed some minor stuff.  I've yet to give it a spin so I'll do that and then approve if I don't find anything to comment on.

grondo commented:
Thanks! I've addressed the comments so far and will force-push the fixups if that's ok.

I'd like to try spinning up an instance configured with prolog/epilog support on fluke. I'll let you know how that goes.

garlick reviewed (approved):
OK, LGTM

grondo commented:
Setting MWP, we can get feedback on this and incorporate fixes for the next release.
